File: app/src/main/java/com/zionhuang/music/ui/fragments/SuggestionFragment.java
package com.zionhuang.music.ui.fragments;

import android.app.SearchManager;
import android.content.Context;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.jakewharton.rxbinding4.appcompat.RxSearchView;
import com.zionhuang.music.R;
import com.zionhuang.music.ui.adapters.SearchSuggestionAdapter;
import com.zionhuang.music.ui.fragments.SuggestionFragmentDirections.ActionSuggestionFragmentToSearchResultFragment;
import com.zionhuang.music.ui.widgets.RecyclerViewClickManager;
import com.zionhuang.music.viewmodels.SuggestionViewModel;

import java.util.concurrent.TimeUnit;

import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;

import static autodispose2.AutoDispose.autoDisposable;
import static com.zionhuang.music.ui.fragments.SuggestionFragmentDirections.actionSuggestionFragmentToSearchResultFragment;

public class SuggestionFragment extends BaseFragment {
    private static final String TAG = "SearchSuggestionFragment";
    private SuggestionViewModel mViewModel;
    private SearchView mSearchView;

    @Override
    protected int layoutId() {
        return R.layout.fragment_search_suggestion;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mViewModel = new ViewModelProvider(this).get(SuggestionViewModel.class);

        RecyclerView recyclerView = findViewById(R.id.suggestion_recycler_view);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        SearchSuggestionAdapter adapter = new SearchSuggestionAdapter(mViewModel);
        recyclerView.setAdapter(adapter);
        RecyclerViewClickManager.setup(recyclerView, (i, v) -> search(adapter.getQueryByPosition(i)), null);

        setupSearchView();
        mViewModel.onFillQuery().observe(getViewLifecycleOwner(), q -> mSearchView.setQuery(q, false));
        mViewModel.getQuery().observe(getViewLifecycleOwner(), mViewModel::fetchSuggestions);
        mViewModel.getSuggestions().observe(getViewLifecycleOwner(), adapter::setDataSet);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_search_view, menu);
        MenuItem searchItem = menu.findItem(R.id.search_view);
        searchItem.setActionView(mSearchView);
    }

    private void setupSearchView() {
        mSearchView = new SearchView(requireContext());
        mSearchView.setIconified(false);
        EditText searchEditText = mSearchView.findViewById(androidx.appcompat.R.id.search_src_text);
        searchEditText.setPadding(0, 2, 0, 2);
        SearchManager searchManager = (SearchManager) requireContext().getSystemService(Context.SEARCH_SERVICE);
        if (searchManager != null) {
            mSearchView.setSearchableInfo(searchManager.getSearchableInfo(requireActivity().getComponentName()));
        }
        mSearchView.setSubmitButtonEnabled(false);
        mSearchView.setMaxWidth(Integer.MAX_VALUE);
        mSearchView.setOnCloseListener(() -> true);
        RxSearchView.queryTextChangeEvents(mSearchView)
                .debounce(100, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .to(autoDisposable(AndroidLifecycleScopeProvider.from(this)))
                .subscribe(e -> {
                    if (e.isSubmitted()) {
                        search(e.getQueryText().toString());
                    } else {
                        mViewModel.setQuery(e.getQueryText().toString());
                    }
                });
        mSearchView.setQuery(mViewModel.getQuery().getValue(), false);
    }

    private void search(String query) {
        mSearchView.clearFocus();
        ActionSuggestionFragmentToSearchResultFragment action = actionSuggestionFragmentToSearchResultFragment(query);
        NavHostFragment.findNavController(this).navigate(action);
    }
}

File: app/src/main/java/com/zionhuang/music/ui/fragments/SuggestionFragment.kt
package com.zionhuang.music.ui.fragments

import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.widget.EditText
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.zionhuang.music.R
import com.zionhuang.music.databinding.FragmentSearchSuggestionBinding
import com.zionhuang.music.extensions.addOnClickListener
import com.zionhuang.music.extensions.getQueryTextChangeFlow
import com.zionhuang.music.ui.adapters.SearchSuggestionAdapter
import com.zionhuang.music.ui.fragments.base.MainFragment
import com.zionhuang.music.viewmodels.SuggestionViewModel
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.launch
import java.util.concurrent.TimeUnit
import kotlin.time.ExperimentalTime
import kotlin.time.toDuration

class SuggestionFragment : MainFragment<FragmentSearchSuggestionBinding>() {
    companion object {
        private const val TAG = "SearchSuggestionFragment"
    }

    private val viewModel by viewModels<SuggestionViewModel>()
    private lateinit var searchView: SearchView

    @FlowPreview
    @ExperimentalTime
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val suggestionAdapter = SearchSuggestionAdapter(viewModel)
        binding.recyclerView.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(requireContext())
            adapter = suggestionAdapter
            addOnClickListener { pos, _ ->
                search(suggestionAdapter.getQueryByPosition(pos))
            }
        }
        setupSearchView()
        viewModel.apply {
            onFillQuery.observe(viewLifecycleOwner, { query -> searchView.setQuery(query, false) })
            query.observe(viewLifecycleOwner, { query -> viewModel.fetchSuggestions(query) })
            suggestions.observe(viewLifecycleOwner, { dataSet -> suggestionAdapter.setDataSet(dataSet) })
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_search_view, menu)
        menu.findItem(R.id.search_view).actionView = searchView
    }

    @FlowPreview
    @ExperimentalTime
    private fun setupSearchView() {
        val searchManager = requireContext().getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView = SearchView(requireContext()).apply {
            isIconified = false
            findViewById<EditText>(androidx.appcompat.R.id.search_src_text)?.setPadding(0, 2, 0, 2)
            setSearchableInfo(searchManager.getSearchableInfo(requireActivity().componentName))
            isSubmitButtonEnabled = false
            maxWidth = Int.MAX_VALUE
            setOnCloseListener { true }
            viewLifecycleOwner.lifecycleScope.launch {
                getQueryTextChangeFlow()
                        .debounce(100.toDuration(TimeUnit.MILLISECONDS))
                        .collect { e ->
                            if (e.isSubmitted) {
                                search(e.query.orEmpty())
                            } else {
                                viewModel.setQuery(e.query)
                            }
                        }
            }
            setQuery(viewModel.query.value, false)
        }
    }

    private fun search(query: String) {
        searchView.clearFocus()
        val action = SuggestionFragmentDirections.actionSuggestionFragmentToSearchResultFragment(query)
        NavHostFragment.findNavController(this).navigate(action)
    }
}